linera-core/src/worker.rs

```diff
@@ -761,85 +824,131 @@ where
+            height = %certificate.inner().height(),
+        ))]
     pub async fn handle_certificate(
         &self,
         certificate: Certificate,
         blobs: Vec<Blob>,
         notify_when_messages_are_delivered: Option<oneshot::Sender<()>>,
     ) -> Result<(ChainInfoResponse, NetworkActions), WorkerError> {
-        trace!("{} <-- {:?}", self.nickname, certificate);
-
-        #[cfg(with_metrics)]
-        let (round, log_str, mut confirmed_transactions, mut duplicated) = (
-            certificate.round,
-            certificate.inner().to_log_str(),
-            0u64,
-            false,
-        );
-
-        let (info, actions) = match certificate.inner() {
-            CertificateValue::ValidatedBlock { .. } => {
-                // Confirm the validated block.
-                // Note: This conversion panics if `certificate` is not a validated block certificate.
-                let validated_block_certificate: ValidatedBlockCertificate = certificate.into();
-                let validation_outcomes = self
-                    .process_validated_block(validated_block_certificate, &blobs)
-                    .await?;
-                #[cfg(with_metrics)]
-                {
-                    duplicated = validation_outcomes.2;
-                }
-                let (info, actions, _) = validation_outcomes;
-                (info, actions)
-            }
-            CertificateValue::ConfirmedBlock(_confirmed) => {
-                #[cfg(with_metrics)]
-                {
-                    confirmed_transactions = (_confirmed.inner().block.incoming_bundles.len()
-                        + _confirmed.inner().block.operations.len())
-                        as u64;
-                }
-                let confirmed_block_certificate: ConfirmedBlockCertificate =
-                    certificate.try_into().unwrap();
-                // Execute the confirmed block.
-                self.process_confirmed_block(
-                    confirmed_block_certificate,
-                    &blobs,
+        match certificate.inner() {
+            CertificateValue::ConfirmedBlock(_) => {
+                let certificate = ConfirmedBlockCertificate::try_from(certificate).unwrap();
+                self.handle_confirmed_certificate(
+                    certificate,
+                    blobs,
                     notify_when_messages_are_delivered,
                 )
-                .await?
+                .await
             }
-            CertificateValue::Timeout { .. } => {
-                // Handle the leader timeout.
-                // Note: This conversion panics if `certificate` is not a timeout certificate
-                // but we just checked that it is.
-                let timeout_certificate = certificate.into();
-                self.process_timeout(timeout_certificate).await?
+            CertificateValue::ValidatedBlock(_) => {
+                let certificate = ValidatedBlockCertificate::from(certificate);
```

Comment:
For `ConfirmedBlock` we use `try_from(…).unwrap()`. Would be good to be consistent. (Same for `Timeout` below.)

Is the plan to replace `Certificate` with an enum that contains the different types of `GenericCertificate`s? That would make the `unwrap`s here unnecessary.

Reply:
Yes, the plan is to put a `typ` field in the `api::Certificate` definition (protobuf).

Reply:
but I'm trying different approaches.